src/helics/shared_api_library/MessageFederateExport.cpp

```diff
@@ -448,23 +497,111 @@ helics_message helicsFederateGetMessage (helics_federate fed)
-        mess.original_source = fedObj->lastMessage->original_source.c_str ();
-        mess.original_dest = fedObj->lastMessage->original_dest.c_str ();
-        mess.source = fedObj->lastMessage->source.c_str ();
-        mess.time = static_cast<helics_time> (fedObj->lastMessage->time);
+        mess.data = message->data.data ();
+        mess.dest = message->dest.c_str ();
+        mess.length = message->data.size ();
+        mess.original_source = message->original_source.c_str ();
+        mess.source = message->source.c_str ();
+        mess.original_dest = message->original_dest.c_str ();
+        mess.time = static_cast<helics_time> (message->time);
+        fedObj->messages.push_back (std::move (message));
         return mess;
     }
     return emptyMessage ();
 }
 
+helics_message_object helicsFederateGetMessageObject (helics_federate fed)
+{
+    auto mFed = getMessageFed (fed, nullptr);
+    if (mFed == nullptr)
+    {
+        return nullptr;
+    }
+
+    auto fedObj = helics::getFedObject (fed, nullptr);
+
+    auto message = mFed->getMessage ();
+
+    if (message)
+    {
+        helics_message_object mess = message.get ();
+        fedObj->messages.push_back (std::move (message));
+        return mess;
+    }
+    return nullptr;
+}
+
+helics_message_object helicsFederateCreateMessageObject (helics_federate fed, helics_error *err)
+{
+    auto fedObj = helics::getFedObject (fed, err);
+    if (fedObj == nullptr)
+    {
+        return nullptr;
+    }
+    auto messptr = std::make_unique<helics::Message> ();
+    helics_message_object mess = messptr.get ();
+    fedObj->messages.push_back (std::move (messptr));
+    return mess;
+}
+
+void helicsFederateClearMessages (helics_federate fed)
+{
+    auto fedObj = helics::getFedObject (fed, nullptr);
+    if (fedObj == nullptr)
+    {
+        return;
+    }
+    fedObj->messages.clear ();
+}
+
+/** clear all message from an endpoint
+@param endpoint  the endpoint object to operate on
+*/
+void helicsEndpointClearMessages (helics_endpoint endpoint)
+{
+    auto endObj = verifyEndpoint (endpoint, nullptr);
+    if (endObj == nullptr)
+    {
+        return;
+    }
+    endObj->messages.clear ();
+}
+
+helics_message_object helicsFederateGetLastMessage (helics_federate fed)
+{
+    auto fedObj = helics::getFedObject (fed, nullptr);
+    if (fedObj == nullptr)
+    {
+        return nullptr;
+    }
+    if (!fedObj->messages.empty ())
+    {
+        helics_message_object mess = fedObj->messages.back ().get ();
+        return mess;
+    }
+    return nullptr;
+}
+
+helics_message_object helicsEndpointGetLastMessage (helics_endpoint endpoint)
+{
+    auto endObj = verifyEndpoint (endpoint, nullptr);
+    if (endObj == nullptr)
+    {
+        return nullptr;
+    }
+    if (!endObj->messages.empty ())
+    {
+        helics_message_object mess = endObj->messages.back ().get ();
+        return mess;
+    }
+    return nullptr;
+}
+
```

Comment:
`Last` message is ambiguous. Does this refer to oldest message? Users are most likely going to want to use the current time as a reference, so last and first in this context is not clear.

Reply:
hmmm,   how about `getRecentMessage` or `getMostRecentMessage`   I could also just drop this function it isn't required.  It is there for convenience since it would be easy to go back and get a message again now.